OpenRA.Mods.Common/Traits/Render/WithIdleOverlay.cs

```diff
@@ -47,13 +48,34 @@ public IEnumerable<IActorPreview> RenderPreviewSprites(ActorPreviewInitializer i
 				p = init.WorldRenderer.Palette(Palette);
 
 			var body = init.Actor.TraitInfo<BodyOrientationInfo>();
-			var facing = init.Contains<FacingInit>() ? init.Get<FacingInit, int>() : 0;
-			var anim = new Animation(init.World, image, () => facing);
-			anim.PlayRepeating(RenderSprites.NormalizeSequence(anim, init.GetDamageState(), Sequence));
 
-			var orientation = body.QuantizeOrientation(new WRot(WAngle.Zero, WAngle.Zero, WAngle.FromFacing(facing)), facings);
-			var offset = body.LocalToWorld(Offset.Rotate(orientation));
-			yield return new SpriteActorPreview(anim, offset, offset.Y + offset.Z + 1, p, rs.Scale);
+			if (init.Contains<DynamicFacingInit>())
+			{
+				var facing = init.Get<DynamicFacingInit, Func<int>>();
+				var anim = new Animation(init.World, image, facing);
+				anim.PlayRepeating(RenderSprites.NormalizeSequence(anim, init.GetDamageState(), Sequence));
+
+				Func<WRot> orientation = () => body.QuantizeOrientation(WRot.FromFacing(facing()), facings);
+				Func<WVec> offset = () => body.LocalToWorld(Offset.Rotate(orientation()));
+				Func<int> zOffset = () =>
+				{
+					var tmpOffset = offset();
+					return tmpOffset.Y + tmpOffset.Z + 1;
+				};
+
+				yield return new SpriteActorPreview(anim, offset, zOffset, p, rs.Scale);
+			}
+			else
+			{
+				var facing = init.Contains<FacingInit>() ? init.Get<FacingInit, int>() : 0;
+				var anim = new Animation(init.World, image, () => facing);
+				anim.PlayRepeating(RenderSprites.NormalizeSequence(anim, init.GetDamageState(), Sequence));
+
+				var orientation = body.QuantizeOrientation(WRot.FromFacing(facing), facings);
+				var offset = body.LocalToWorld(Offset.Rotate(orientation));
+				var zOffset = offset.Y + offset.Z + 1;
+				yield return new SpriteActorPreview(anim, () => offset, () => zOffset, p, rs.Scale);
+			}
```

Comment:
Is there a way to extract this code into a common method somewhere? It's far from trivial and used in several places.